### PR TITLE
Minor corrections

### PR DESCRIPTION
**Pull request by Commodore64user**
### what's new:

* minor corrections to issues raised by @poire-z here #12162

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12372)

### CODE REVIEW

File: frontend/ui/widget/inputtext.lua [hunk 1]
```diff
@@ -655,6 +655,8 @@ function InputText:onKeyPress(key)
     else
         handled = false
     end
+    -- This primarily targets Kindle. When a virtual keyboard is shown on screen, mod+dpad allows controlling the cursor, as dpad alone
```
Comment (Commodore64user) [inline frontend/ui/widget/inputtext.lua:658]:
```suggestion
    -- This primarily targets Kindle. When a virtual keyboard is shown on screen, mod+dpad allows controlling the cursor, as dpad alone
```

**Comment by Commodore64user** [comment]
I would like to politely request permission to make an illegal move

to turn this:

https://github.com/koreader/koreader/blob/2403087f681c245f7f4f35c13d240d93594ef174/frontend/apps/reader/modules/readerconfig.lua#L36-L40

into this:

```lua
function ReaderConfig:registerKeyEvents()
    if Device:hasKeys() then
        self.key_events.ShowConfigMenu = { { { "Press", "AA" } } }
        if Device:hasSymKey() then
            self.key_events.ShowConfigMenu = { { "AA" } }
        end
    end
end
```

**Comment by Frenzie** [comment]
What would you do with Press?

**Comment by Commodore64user** [comment]
currently, three buttons call up the bottom menu on k3 <kbd>press</kbd>, <kbd>enter</kbd> and <kbd>AA</kbd>. that bothers me very much.

> What would you do with Press?

mmm, nothing? refresh the screen?

**Review by poire-z — APPROVED** [review]

**Comment by Frenzie** [comment]
I don't understand how it could bother you to have a super convenient way to bring up the menu while navigating, but everything you wrote there screams user patch imho.

**Comment by Commodore64user** [comment]
> I don't understand how it could bother you to have a super convenient way to bring up the menu while navigating, but everything you wrote there screams user patch imho.

For a start there's three buttons, and two of them only do half the work, by that i mean they open the bottom menu but cannot close it (because now they have to work as select). Aa is only one that satisfies both opening and closing just fine.

Is not really any more convenient as far as i am concerned, as accidental triggers seem to happen (to me at least) sometimes. I understand why is needed on k4 (unless...) but on the other kindles is just not.

**Comment by Frenzie** [comment]
> they open the bottom menu but cannot close it (because now they have to work as select)

Which is fantastic. You press, press, press exactly where you want to go. If you don't like it, user patch, don't mess things up for other people.

**Comment by Commodore64user** [comment]
> > they open the bottom menu but cannot close it (because now they have to work as select)
> 
> Which is fantastic. You press, press, press exactly where you want to go. If you don't like it, user patch, don't mess things up for other people.

It isn't fantastic, particularly when is something you only visit maybe once per book. I'm not happy with the outcome, but I was kind of expecting this answer, so. ;-) 

I tried building one but ended up getting that "no engine" error or whatever.

**Comment by Commodore64user** [comment]
@Frenzie feel free to merge this. nothing else is happening here

**Comment by Commodore64user** [comment]
I think I found a way to solve two problems in a reasonable way @Frenzie. Because I don't want it to look like I am trying to scurry around this previous discussion, I want to be upfront about it. 

I do want to use "Press" for refreshing the screen albeit, I won't force it on anyone so instead, a user can decide to do just that. On k4, the bottom menu should appear as an extra icon at the top of the screen and on <k3 well, nothing as there is a dedicated <kbd>Aa</kbd> that would handle it.

I have this so far https://github.com/koreader/koreader/compare/master...Commodore64user:KOReader_fork:refresh-screen if you want to have a look, however I still need to figure out why the `FullRefresh` event won't actually refresh the screen...

**Comment by Frenzie** [comment]
I suppose, though I have some fear that might be entering technical debt territory.

**Comment by Commodore64user** [comment]
the other possible thing is to do the switcheroo. Basically have mod+press do the screen refresh (so everyone gets it either way) and move the current `add location to stack` to this setting. So it would be something like `add location to stack with press key`. And the point to doing this, is that eventually this would change to `page browser with press key`. essentially a placeholder for the time being.

That seems a lot less like a hack.

**Comment by Commodore64user** [comment]
okay, I bit the bullet and, have been trying to get a "gestures" like plugin for shortcuts.

I've gotten the UI side of it, but still needs more stuff and I could use some help here, https://github.com/koreader/koreader/compare/master...Commodore64user:KOReader_fork:shortcuts?expand=1

any chance @poire-z or @NiLuJe are bored looking at flies passing by?

**Comment by poire-z** [comment]
Not bored enough, sorry.
But you may want to name all this more specifically than just "shortcuts", which can have other meaning (folder shortcuts, book shorcuts...)
KeyShortcuts, KeyMaps...

**Comment by Frenzie** [comment]
What kind of stuff? (I don't really have time atm though)

Agreed on the shortcuts thing. Keyboard shortcuts?

**Comment by Commodore64user** [comment]
> What kind of stuff? (I don't really have time atm though)

well for one, matching the shortcuts with the actual key presses.

**Comment by Commodore64user** [comment]
another thing, for some reason the shortcuts that use screenkb or shift show up fine but the ones with alt appear as `%1` instead of using their names. In any case, it would be very helpful if someone could have a look at the whole thing because I won't certainly be able to continue without help. It's a kinda, pick a spade and start digging kind of situation here atm.

https://github.com/koreader/koreader/compare/master...Commodore64user:KOReader_fork:shortcuts?expand=1

**Comment by poire-z** [comment]
https://github.com/Commodore64user/KOReader_fork/blob/fe0d49a52a5d175b5812af025ac187db684d0ada/plugins/hotkeyshortcuts.koplugin/main.lua#L27-L29
is a table/associative array (key => value).
Tables can also be a plain array (1,2,3,4 => values)
(Lua tables can be both, have integer indexes, and string keys, but it's not a good idea to mix them when not needed for some reasons.)
table.insert() only makes sense with plain array, to append values.

https://github.com/Commodore64user/KOReader_fork/blob/fe0d49a52a5d175b5812af025ac187db684d0ada/plugins/hotkeyshortcuts.koplugin/main.lua#L42-L45
is wrong: you are inserting at index 1 another associative array.

You probably want to use:
https://github.com/koreader/koreader/blob/00984ba6bc7827a80e3e33f7ec4113696fbe5116/frontend/util.lua#L366-L374

----
https://github.com/Commodore64user/KOReader_fork/blob/fe0d49a52a5d175b5812af025ac187db684d0ada/plugins/hotkeyshortcuts.koplugin/main.lua#L92-L95
I think here, you are just on each line overwriting the thing you set on previous line.
May be https://github.com/koreader/koreader/issues/9059#issuecomment-2323041900 can help.

**Comment by Commodore64user** [comment]
Thank you @poire-z, i'll have a look at merging the arrays later. Regarding the last bot of your comment though, yes i might be but that is not the real issue atm (they all just temporarily look the same, i know how to write them ;) ) the real issue is that i am not sure how to write the function that plugs both the key_events with the array in `hotkeyshortcuts_path`

**Comment by poire-z** [comment]
> but that is not the real issue atm 

Looks like it was what you were asking :/

> the real issue is that i am not sure how to write the function that plugs both the key_events with the array in hotkeyshortcuts_path

I guess (like in gestures.koplugin that you copied it from) that `function HotKeyShortcuts:onFlushSettings()` takes care of that. You just update self.settings.data (a datastructure of your choice) and this is called when leaving reader/fm and stuff is written to the file.

> Btw is the `if self.ui ~= nil then ` correct? Is supposed to be used only whilst inside reader

Probably not, all plugins, even in FM, gets a self.ui = reader or filemanager.
Not sure how other modules check that, may be if self.ui and self.ui.document.
`self.is_docless` might be just be doing that.

**Comment by Commodore64user** [comment]
> Looks like it was what you were asking :/

oops, sorry i was outside, so couldn’t read it properly then.

> I guess (like in gestures.koplugin that you copied it from) that `function HotKeyShortcuts:onFlushSettings()` takes care of that. You just update self.settings.data (a datastructure of your choice) and this is called when leaving reader/fm and stuff is written to the file.

i am still not sure about the whole thing though, could you please provide an example relative to this particular case?

**Comment by poire-z** [comment]
I'm still not sure about what you're asking. Anyway, the existing example is gestures.koplugin: follow what its self.settings.data is and how it gets updated when the user change things - I haven't followed, so don't ask me. I'm just telling you that the saving/writting is auto taken care of by this HotKeyShortcuts:onFlushSettings() (which is called by other parts of the code, don't worry about that).

**Comment by Commodore64user** [comment]
No, what i am asking is how to match the event selected by user that has been saved to that file (or from the default) to the key-mapping.

essentially when i press mod+up i am triggering whatever event was associated to that. That is what i am missing.

**Comment by poire-z** [comment]
> No, what i am asking is how to match

I didn't read you asking that...
Which is indeed the toughest part of your project - and no obvious answer.
Try to undestand how InputContainer works, how keys are registered and handled.

gestures.koplugin > InputContainer:registerTouchZones / unRegisterTouchZones was how it's handled for touch stuff.
You have to figure out how to do the same for keys (managing the lifecycle of your key mapping changes).

**Comment by Frenzie** [comment]
If you make it a modal container at the relevant time (also see the gesture creator) it should capture all the events.

**Comment by Commodore64user** [comment]
I think I finally figured it out!!! now I have a question regarding memory and performance.

how does having lot of key_events ready to fire at any time affect it? should I gate lock some of those behind a user toggle? for example all the alt+alphabet ones.

**Comment by Frenzie** [comment]
But you already have a lot ready to fire?

**Comment by Commodore64user** [comment]
By default everything would be `nil`, only what is currently in place would be set as default. So all the alt+alphabet would be empty apart from the wiki and search ones for example.

**Comment by Frenzie** [comment]
Right, I just mean the performance is unlikely to be noticably different.

**Comment by Commodore64user** [comment]
> Right, I just mean the performance is unlikely to be noticably different.

understood. I'll plough ahead then.

I realised that inverting page turn buttons, will also invert key_mappins, which can make you scratch your head for a while before you realise what is going on. Any advise on how to handle that? I tried that trick with the connect keyboard thing but that didn't re-registered the key events. `HotKeyShortcuts.onSwapPageTurnButtons = HotKeyShortcuts.registerKeyEvents`

**Comment by Commodore64user** [comment]
how can say `if Device:hasScreenKB() or Device:hasSymKey() and SHORTCUTS PLUGIN IS ON then`?

**Comment by poire-z** [comment]
Depends on where.
`not self.disabled` in the plugin itself
`self.ui.hotkeyshortcuts` elsewhere to know if the plugin is loaded
`self.ui.hotkeyshortcuts and not self.ui.hotkeyshortcuts.disabled` elsewhere to know it's loaded and not disabled.

**Comment by Commodore64user** [comment]
thanks!

and what about being here:
`./luajit: frontend/ui/elements/physical_buttons.lua:67: attempt to index global 'self' (a nil value)`

**Comment by Commodore64user** [comment]
this one works great

```lua
function ReaderToc:registerKeyEvents()
    local hotkey_plugin = self.ui.shortcuts and not self.ui.shortcuts.disabled
    if Device:hasScreenKB() and hotkey_plugin then
        self.key_events.ShowToc = { { "ScreenKB", "Up" } }
    elseif Device:hasKeyboard() and hotkey_plugin then
        self.key_events.ShowToc = { { "T" }, { "Shift", "Up" } }
    end
end
```

**Comment by poire-z** [comment]
> and what about being here:
> `./luajit: frontend/ui/elements/physical_buttons.lua:67: attempt to index global 'self' (a nil value)`

Well, no idea. This one is require()'d, so loaded once, and doesn't have much visbility/access to other stuff.
So, you may need to create a setting, or plug something into Device.
But we'd also want your plugin to not intrude too uglily into core stuff. So, think.

**Comment by Commodore64user** [comment]
That https://github.com/koreader/koreader/pull/12372?notification_referrer_id=NT_kwDOBdFQh7QxMTk5Mzc0MTc4NTo5NzYwMzcxOQ#issuecomment-2336482345 constitute too uglily into core? That is how i was thinking about solving the issue of which events to load

**Comment by poire-z** [comment]
I don't really know (and don't really want to become more involved), but given that it wouldn't work in the case of physical_button, may be it needs another solution (like the plugin setting up a `Device:hasConfigurableKeyShortcuts() return true end`, which would then be available everywhere). 
Also, your plugin (if it stays a plugin) may be disabled, and you'd still want some "default" keymap to work. The snippet in your comment seems like it wouldn't allow showToc if the plugin is disabled.

**Comment by Commodore64user** [comment]
> and don't really want to become more involved 

why do you always say that?, its kinda hurtful at this point. ;)

> if it stays a plugin

Is there another option? I suppose it could be its own `ReaderShortcuts`

> The snippet in your comment seems like it wouldn't allow showToc if the plugin is disabled.

Yes, that is true. I was aware of it and obviously would make it so it does kick in when is disabled.

bu i am curious as to what @Frenzie has in mind as well.

**Comment by Frenzie** [comment]
In mind about what?

**Comment by Commodore64user** [comment]
> In mind about what?

Anything you’d like to share really, ;). More specifically though https://github.com/koreader/koreader/pull/12372#issuecomment-233662800 and surrounding comment area

**Comment by Frenzie** [comment]
The link doesn't really seem to go anywhere oddly enough.

**Comment by Commodore64user** [comment]
https://github.com/koreader/koreader/pull/12372#issuecomment-2336628000

**Comment by Frenzie** [comment]
If keyboard shortcuts can't be done sanely as a plugin they should either be part of the core or the core should be adapted, but as written I don't care for that. But my main thoughts are what @yparitcher wrote in some related issue.

**Comment by Commodore64user** [comment]
Sigh, i just wanted the k3 not to trigger the bottom menu with <kbd>press</kbd> ;), okay partly true.

are you saying that, the current form of the plugin is NOT done sanely?

**Comment by Frenzie** [comment]
If it necessitates putting an `if plugin` in every file there's definitely something not quite right yet. But anything I want to say about that has already been mentioned in https://github.com/koreader/koreader/pull/12484#issuecomment-2336535702 and following.

**Comment by Commodore64user** [comment]
> If it necessitates putting an `if plugin` in every file there's definitely something not quite right yet. But anything I want to say about that has already been mentioned in [#12484 (comment)](https://github.com/koreader/koreader/pull/12484#issuecomment-2336535702) and following.

okay, the priority is that all shortcuts in the plugin take precedence, always, no exceptions. I have already, by design made sure that there are no issues there (I mean I've implemented everything so I should know), for example, one cannot use <kbd>mod</kbd> + <kbd>press</kbd> (which is long-press or hold) in FM. so how could I write that directly in the plugin?

**Comment by Commodore64user** [comment]
@poire-z could I bother you one last time here, How could I make this plugin a user patch or something I don't have to worry about when updating to new KOReader stable versions? I suppose I would need a patch that overrides the `registerEvents` on the offending `ReaderFiles` and `FileManager` ones, could that be one single file with alll the [new] functions there?

**Comment by poire-z** [comment]
I think allowing customizing key presses with user patches would be a lot easier than the whole plugin idea you had or been discussing in the other PR (the UI would be complex, and the target audience is limited, being only users with old Kindle as no new device gets keys).

And I think it's just easy, as most of the :registerKeyEvents() are quite small: just replace them, and it's dosable in a single user patch, ie.:

```lua
local ReaderToc = require("apps/reader/modules/readertoc")
ReaderToc.registerKeyEvents = function(self)
        self.key_events.ShowToc = { { "ScreenKB", "Left" } }
end

local ReaderDictionary= require("apps/reader/modules/readerdictionary")
ReaderDictionary.registerKeyEvents = function(self)
        self.key_events.ShowDictionaryLookup = { { "ScreenKB", "Down" } }
end
...
```

**Comment by Commodore64user** [comment]
Thanks for the patch, still, is there a way to run plugin regardless? Btw it wasn't meant just for old kindles, it was also for anyone with a keyboard.

It would be simultaneously running  both the patch and the plugin in case i am not being clear enough.

**Comment by poire-z** [comment]
> It would be simultaneously running both the patch and the plugin in case i am not being clear enough.

Not sure you are :)
Just put and enable your plugin, and put a user patch replacing what needs to be replaced in core. ?

**Comment by Commodore64user** [comment]
yes, I just want to run it myself given that now the expectations seem to have gone way up, and therefore won't be accepted. my concerns are thus, 1. would new updates disrupt the existence of the plugin?, 2. what about adding it to the main menu?

**Comment by poire-z** [comment]
If you put a plugin in our plugins/ alongside all the others, as long as we don't ship a plugin with the same dirname, it will persist. And anything menu it adds will be put in the plugin submenu, I think (or there is a `hint/order =` to tells a menu item to be put in another menu.

**Comment by Commodore64user** [comment]
but wouldn't I also need to add it to `reader_menu_order.lua` and `filemanager_menu_order.lua` and technically speaking it also needs to add 
```lua
if self.ui.hotkeyshortcuts then -- search and update the profile action in assigned keyboard shortcuts
   self.ui.hotkeyshortcuts:updateProfiles(action_old_name, action_new_name)
end
```
in the profiles koplugin

**Comment by poire-z** [comment]
> but wouldn't I also need to add it to `reader_menu_order.lua` and `filemanager_menu_order.lua`

Plugins can use a `sorting_hint = ("more_tools")` to tell in which menu/submenu they'd like their item to be put.

> technically speaking it also needs to add in the profiles koplugin

You can user-patch plugins I think (may depend on the amount of tweak you need), I added something to make that easier, see https://github.com/koreader/koreader/pull/11838#issuecomment-2124493747.

**Comment by Commodore64user** [comment]
> Plugins can use a sorting_hint = ("more_tools") to tell in which menu/submenu they'd like their item to be put.

I see, sorry I didn't understand it before.

regarding the user patch for the `registerKeyEvents`, is that a number 0 or 1? and the user-patch plugin, I am not really sure I understand what to do. I just need to add those three lines seen here https://github.com/koreader/koreader/pull/12372#issuecomment-2348534395 to https://github.com/koreader/koreader/blob/57ceaf3f07eb54ca0ee14c2a5a61dd108d062e97/plugins/profiles.koplugin/main.lua#L392-L394

**Comment by poire-z** [comment]
> regarding the user patch for the registerKeyEvents, is that a number 0 or 1? 

You mean the number at the start of the filename ? `2-`, like for most user patches.

> and the user-patch plugin, I am not really sure I understand what to do. I just need to add those three lines seen here 

Untested:

```lua
local userpatch = require("userpatch")
userpatch.registerPatchPluginFunc("profiles", function(plugin)
    local orig_updateProfiles = plugin.updateProfiles
    plugin.updateProfiles = function(self, action_old_name, action_new_name)
        orig_updateProfiles(self, action_old_name, action_new_name)
        if self.ui.gestures then -- search and update the profile action in assigned gestures
            self.ui.gestures:updateProfiles(action_old_name, action_new_name)
        end
    end
end)
```

**Comment by Commodore64user** [comment]
> You mean the number at the start of the filename ? 

yes that was it. I'll try the other patch. Thank you!!!!

so I am running the plugin, everything seems fine except for one little thing. Every time I open the menu, a new pair of the settings introduced by the plugin are added to the menu, but I don't know why



so that is the third time, opening it, and as you can see, three sets of options. https://github.com/koreader/koreader/compare/master...Commodore64user:KOReader_fork:hotkeyshortcuts?expand=1

**Comment by poire-z** [comment]
Dunno either.
Does it happen with and/or without the sorting_hint ?
Does it happen with and/or without the user patch plugin stuff ?

**Comment by Commodore64user** [comment]
> Does it happen with and/or without the user patch plugin stuff ?

I am only building it now, so not this.

> Does it happen with and/or without the sorting_hint ?

without it, it does that weird thing where it puts the settings in the first icon (page and bookmark) as `NEW: Use press...` and `NEW: Shortcuts` since I am no longer telling it where to go, but it doesn't seem to multiply them as you revisit.

**Comment by poire-z** [comment]
> without it, it does that weird thing where it puts the settings in the first icon (page and bookmark) as NEW: Use press... and NEW: Shortcuts since I am no longer telling it where to go

That's by design and normal.

> but it doesn't seem to multiply them as you revisit.

So, may be a bug with sorting_hint, @Frenzie ?

**Comment by Commodore64user** [comment]
note about replication, open in FM first, open a book and then open the menu in reader (setting will be duplicated), close reader and open menu in FM (settings triplicated, as in screenshot). 

If you only do it, in either FM or reader, subsequent attempts do not duplicate settings.

**Comment by Frenzie** [comment]
> So, may be a bug with sorting_hint

I don't know what that means. Not multiplying sounds like a good thing though.

**Comment by poire-z** [comment]
I tried replicating the behaviour you describe by hacking the autoturn.koplugin, but couldn't get its menu item duplicated :/

**Comment by poire-z** [comment]
I don't think this could be the culprit, but `sorting_hint = ("physical_buttons_setup")` doesn't need any parens.
Also try with adding only one of your two menu items (in case multiple identical sorting_hints causes the issue).

**Comment by Commodore64user** [comment]
> I don't think this could be the culprit, but `sorting_hint = ("physical_buttons_setup")` doesn't need any parens. Also try with adding only one of your two menu items (in case multiple identical sorting_hints causes the issue).

not it either,


> I tried replicating the behaviour you describe by hacking the autoturn.koplugin, but couldn't get its menu item duplicated :/

why don't you try the shortcuts plugin then? ;) just pushed the latest version I have available.

**Comment by poire-z** [comment]
Did.
It doesn't happen if you use sorting_hints = "navigation".
So, I guess it is because our code uses `self.menu_items.physical_buttons_setup = require("ui/elements/physical_buttons")`, and it is always the cached require'd object that gets re-used and each time got re-added your items.
We probably didn't need to think about that case, before you came in (these were happy times :))

**Comment by Commodore64user** [comment]
so essentially, what you are saying is, this is a: "is not you, it's me" kind of situation ;)

**Comment by poire-z** [comment]
No, it is a "it's you, and it's @Frenzie" kind of situation :)
May be something like this could be added (seems to solve duplication, but not familiar with the whole menusorter stuff, and how this additional loop adds to the already complex processing)?
```diff
--- a/frontend/ui/menusorter.lua
+++ b/frontend/ui/menusorter.lua
@@ -179,6 +179,12 @@ function MenuSorter:sort(item_table, order)
         if sorting_hint then
             local sorting_hint_menu = self:findById(menu_table["KOMenu:menu_buttons"], sorting_hint)
             sorting_hint_menu = sorting_hint_menu.sub_item_table or sorting_hint_menu
+            for i=1, #sorting_hint_menu do
+                if sorting_hint_menu[i].id == v.id then
+                    table.remove(sorting_hint_menu, i)
+                    break
+                end
+            end
             table.insert(sorting_hint_menu, v)
         else
             table.insert(menu_table["KOMenu:menu_buttons"][1], v)
```

**Comment by NiLuJe** [comment]
Or, better yet, just use `dofile` ;).

**Comment by Commodore64user** [comment]
> Or, better yet, just use `dofile` ;). 

Mmm, what does that mean? 🤔

**Comment by Commodore64user** [comment]
I have been running the plugin on both k3 and k4 for the last few days, seems to work absolutely fine, so I ask. Are you interested in re-opening that PR? I can't make the changes that were quickly piled up there. And, would certainly need to hack the keyEvent functions like in https://github.com/koreader/koreader/pull/12372#issuecomment-2336482345, thoughts?

**Comment by Commodore64user** [comment]
Is that a No?

**Comment by Frenzie** [comment]
That was addressed to NiLuJe?

**Comment by NiLuJe** [comment]
Not sure what we're talking about, exactly ^^.

**Comment by Commodore64user** [comment]
It was a question for all of you, in short:

There is shortcuts plugin, it works, is not perfect. I personally can't make it better at this moment in time but would like to make it available to all... it would require some concessions, more precisely, making ugly adjustments to the current key_events functions affected.

**Comment by NiLuJe** [comment]
I'm pretty firmly in the "hell no" camp, then ;).

**Comment by NiLuJe** [comment]
I meant the whole Key framework, pretty much from top to bottom ;).

**Comment by Commodore64user** [comment]
I am with you, that’s why i am saying I would not want to change anything else for it (plugin). It is what it is. The only thing would be to hack the functions with current events like in https://github.com/koreader/koreader/pull/12372#issuecomment-2336482345

unless someone else wanted to provide a function that designates priorities ;)

**Pull request merged**